src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -71,99 +83,280 @@ public long BytesWritten
         {
-            Debug.Assert(count >= 0 && _buffered <= int.MaxValue - count);
+            BytesPending = default;
+            BytesCommitted = default;
+            _memory = default;
 
-            _buffered += count;
-            _buffer = _buffer.Slice(count);
+            _inObject = default;
+            _isNotPrimitive = default;
+            _tokenType = default;
+            _currentDepth = default;
+
+            // Only allocate if the user writes a JSON payload beyond the depth that the _allocationFreeContainer can handle.
+            // This way we avoid allocations in the common, default cases, and allocate lazily.
+            _bitStack = default;
         }
 
         /// <summary>
-        /// Advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// Commits the JSON text written so far which makes it visible to the output destination.
         /// </summary>
-        /// <param name="isFinalBlock">Let's the writer know whether more data will be written. This is used to validate
-        /// that the JSON written so far is structurally valid if no more data is to follow.</param>
-        /// <exception cref="InvalidOperationException">
-        /// Thrown when incomplete JSON has been written and <paramref name="isFinalBlock"/> is true.
-        /// (for example when an open object or array needs to be closed).
-        /// </exception>
-        public void Flush(bool isFinalBlock = true)
+        /// <remarks>
+        /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// In the case of Stream, this writes the data to the stream and flushes it.
+        /// </remarks>
+        public void Flush()
         {
-            if (isFinalBlock && !_writerOptions.SkipValidation && (CurrentDepth != 0 || _tokenType == JsonTokenType.None))
-                ThrowHelper.ThrowInvalidOperationException_DepthNonZeroOrEmptyJson(_currentDepth);
+            FlushHelper();
 
+            if (_stream != null)
+            {
+                FlushHelperStream();
+                _stream.Flush();
+            }
+
+            _memory = default;
+            BytesCommitted += BytesPending;
+            BytesPending = 0;
+        }
+
+        /// <summary>
+        /// Commits any left over JSON text that has not yet been flushed which makes it visible to the output destination.
+        /// </summary>
+        /// <remarks>
+        /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// In the case of Stream, this writes the data to the stream and flushes it.
+        /// </remarks>
+        /// <remarks>
+        /// The <see cref="Utf8JsonWriter"/> can still be re-used after disposing but all its state has been reset.
+        /// </remarks>
+        public void Dispose()
```

Comment:
Should the Utf8JsonWriter become unusable after calls to Dispose? Currently, its a glorified Reset. Should I null the IBW/Stream fields?

Do we want the "dispose stream" semantics here?

Reply:
> Should the Utf8JsonWriter become unusable after calls to Dispose?

Yep.  Ideally the other methods (aside from Dispose or DisposeAsync) throw ObjectDisposedException after a call to Dispose. But the most important thing is to not let Reset wake it up again.

Reply:
> But the most important thing is to not let Reset wake it up again.

That's what I was thinking of guarding against. The current semantics let it be awaken since all it does is reset.